src/utils/str.c

```diff
@@ -10,93 +11,172 @@
 #include <stdlib.h>
 #include <string.h>
 
-#define READ_BLOCK_SIZE (1024)
+#define STR_STACK_SIZE sizeof(str)
+#define STR_FLAGBIT_COUNT 1
+
+// ------------------------ Basic Accessors ------------------------
+
+// the number of spare capacity in the small string representation.
+// also contains the flag bits for detecting small/large strings.
+static inline uint8_t str_smallspare(str const *s) {
+    uint8_t spare;
+    memcpy(&spare, &s->small[STR_STACK_SIZE - 1], 1);
+    return spare;
+}
+
+static inline size_t str_normalcapacity(str const *s) {
```

Comment:
Some comments to these on what happens and WHY would be excellent.

Reply:
Documented the memory layout, and added a bunch of comments throughout that section.